### PR TITLE
fix unstable raft ut

### PR DESCRIPTION
**Pull request by critical27**
See [failed log](https://github.com/vesoft-inc/nebula/pull/2223/checks?check_run_id=863918745).

In some of the ut, RaftPart would be stop and restart. When `restart`, it would add a background task `statusPolling`. And this would make part would start election at wrong time.

In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance. 

BTW, fix StoragePerf tool always write the same edge.

### CODE REVIEW

**Comment by dangleptr** [comment]
> In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.

You mean the electionRequest is sent out after the part has been stopped?

**Comment by critical27** [comment]
> > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> 
> You mean the electionRequest is sent out after the part has been stopped?

Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.

**Comment by dangleptr** [comment]
> > > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> > 
> > 
> > You mean the electionRequest is sent out after the part has been stopped?
> 
> Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.

But after the part stopped,  the statusPolling task will not send vote request anymore.  You have not called `stop` yet?

**Comment by critical27** [comment]
> > > > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> > > 
> > > 
> > > You mean the electionRequest is sent out after the part has been stopped?
> > 
> > 
> > Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.
> 
> But after the part stopped, the statusPolling task will not send vote request anymore. You have not called `stop` yet?

Yeah, we don't stop `statusPolling` when we stop the `RaftPart` for now.

**Comment by critical27** [comment]
image

**Comment by dangleptr** [comment]
> > > > > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> > > > 
> > > > 
> > > > You mean the electionRequest is sent out after the part has been stopped?
> > > 
> > > 
> > > Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.
> > 
> > 
> > But after the part stopped, the statusPolling task will not send vote request anymore. You have not called `stop` yet?
> 
> Yeah, we didn't stop `statusPolling` when we stop the `RaftPart` for now. One reason is that all instance share the same `bgThread_`.

Not exactly,  i think i got the root cause.  It is one ABA problem,  it means after stopping the Part,  the status has been reset to STOPPED.  Now there is still one left task inside your eventLoop,   when the tasking invoking,  you restart it and the status has been reset to RUNNING.

**Comment by critical27** [comment]
> Not exactly, i think i got the root cause. It is one ABA problem, it means after stopping the Part, the status has been reset to STOPPED. Now there is still one left task inside your eventLoop, when the task is invoked, you restart the part and the status has been reset to RUNNING.



> > > > > > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> > > > > 
> > > > > 
> > > > > You mean the electionRequest is sent out after the part has been stopped?
> > > > 
> > > > 
> > > > Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.
> > > 
> > > 
> > > But after the part stopped, the statusPolling task will not send vote request anymore. You have not called `stop` yet?
> > 
> > 
> > Yeah, we didn't stop `statusPolling` when we stop the `RaftPart` for now. One reason is that all instance share the same `bgThread_`.
> 
> Not exactly, i think i got the root cause. It is one ABA problem, it means after stopping the Part, the status has been reset to STOPPED. Now there is still one left task inside your eventLoop, when the task is invoked, you restart the part and the status has been reset to RUNNING.

Yeah exactly. The key point is we need to prevent two thread doing `statusPolling`, otherwise the problem still exists.

**Comment by dangleptr** [comment]
> > Not exactly, i think i got the root cause. It is one ABA problem, it means after stopping the Part, the status has been reset to STOPPED. Now there is still one left task inside your eventLoop, when the task is invoked, you restart the part and the status has been reset to RUNNING.
> 
> > > > > > > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> > > > > > 
> > > > > > 
> > > > > > You mean the electionRequest is sent out after the part has been stopped?
> > > > > 
> > > > > 
> > > > > Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.
> > > > 
> > > > 
> > > > But after the part stopped, the statusPolling task will not send vote request anymore. You have not called `stop` yet?
> > > 
> > > 
> > > Yeah, we didn't stop `statusPolling` when we stop the `RaftPart` for now. One reason is that all instance share the same `bgThread_`.
> > 
> > 
> > Not exactly, i think i got the root cause. It is one ABA problem, it means after stopping the Part, the status has been reset to STOPPED. Now there is still one left task inside your eventLoop, when the task is invoked, you restart the part and the status has been reset to RUNNING.
> 
> Yeah exactly. The key point is we need to prevent two thread doing `statusPolling`, otherwise the problem still exists. Do you have better solution?

I suggest passing current part version into statusPolling,   and check it at first inside the method.
'startTimeMs_' could be reused for this purpose.

**Comment by critical27** [comment]
> startTimeMs_



> > > Not exactly, i think i got the root cause. It is one ABA problem, it means after stopping the Part, the status has been reset to STOPPED. Now there is still one left task inside your eventLoop, when the task is invoked, you restart the part and the status has been reset to RUNNING.
> > 
> > 
> > > > > > > > In the previous failed case, a part send election twice in different thread within a very short period of time, with term 5 and 6, which makes a log of term 5 failed to replicate to followers. However, this should not happen in real instance.
> > > > > > > 
> > > > > > > 
> > > > > > > You mean the electionRequest is sent out after the part has been stopped?
> > > > > > 
> > > > > > 
> > > > > > Actually, multiple thread is trying to start election after it restarted in ut, because we add a `statusPolling` task whenever we call `start`.
> > > > > 
> > > > > 
> > > > > But after the part stopped, the statusPolling task will not send vote request anymore. You have not called `stop` yet?
> > > > 
> > > > 
> > > > Yeah, we didn't stop `statusPolling` when we stop the `RaftPart` for now. One reason is that all instance share the same `bgThread_`.
> > > 
> > > 
> > > Not exactly, i think i got the root cause. It is one ABA problem, it means after stopping the Part, the status has been reset to STOPPED. Now there is still one left task inside your eventLoop, when the task is invoked, you restart the part and the status has been reset to RUNNING.
> > 
> > 
> > Yeah exactly. The key point is we need to prevent two thread doing `statusPolling`, otherwise the problem still exists. Do you have better solution?
> 
> I suggest passing current part version into statusPolling, and check it at first inside the method.
> 'startTimeMs_' could be reused for this purpose.

Good suggestion, fixed.

File: src/kvstore/raftex/RaftPart.cpp [hunk 1]
```diff
@@ -1229,12 +1229,18 @@ void RaftPart::statusPolling() {
     }
     {
         std::lock_guard<std::mutex> g(raftLock_);
+        // If startTime is not same as the time when `statusPolling` is add to event loop,
+        // it means the part has been restarted (it only happens in ut for now), so don't
+        // add another `statusPolling`.
+        if (startTime != startTimeMs_) {
```
Comment (dangleptr) [inline src/kvstore/raftex/RaftPart.cpp:1235]: Maybe we need to check it at first in the method.

**Comment by dangleptr** [comment]
Please double check the failed UTs

**Pull request merged**